File: src/main/java/org/entities/aggressive/RedMajor.java
package org.entities.aggressive;

import org.engine.Main;
import org.entities.Autonomous;
import org.entities.Entity;
import org.entities.SmartRectangle;
import org.graphics.Animator;
import org.graphics.BossBar;
import org.graphics.Graphics;
import org.level.Level;
import org.level.LevelController;
import org.loader.ImageResource;
import org.loader.ResourceHandler;
import org.world.Attack;
import org.world.HeightMap;
import org.world.HeightReturn;
import org.world.World;

public class RedMajor extends Autonomous {
    private final int NORMAL=0,READYING=1,ATTACK=2, DAMAGE =3,DEATH=4;
    private Animator redAnimator;
    private ImageResource redMajor;
    private SmartRectangle hitbox=new SmartRectangle(x,y,width,height);
    private BossBar bossBar=new BossBar(this);
    private boolean doStartReady=true,startFight=false;
    public RedMajor() {
        super(5, 75, 7);
        reset();

    }

    public void setStartFight(boolean startFight) {
        this.startFight = startFight;
    }

    @Override
    public void update() {
        bossBar.update();
        if(state==4){
            if(redAnimator.getDelay()>0)redAnimator.update();
            else health=0;
        }
        if(!startFight||state==4)return;
        HeightReturn heightReturn= HeightMap.onGround(hitbox);
        //Determines movement and knockback
        if(damageTakenFrame==0) {
            if (direction) {
                vX = .15f;
            } else {
                vX = -.15f;
            }
        }else{
            if((direction&&!attackerBehind)||(!direction&&attackerBehind))vX=-.3f;
            else vX=.3f;
            damageTakenFrame--;
        }

        //Calculates the y-value and velocity
        y+=vY;
        vY-= World.getGravity();
        if(heightReturn.isOnGround()&&vY<0){
            y=heightReturn.getGroundLevel();
            vY=0;
        }

        //Calculates the x-value and velocity
        if(state!=READYING&&redAnimator.getDelay()==0) {
            x += vX;
            Level l = LevelController.getCurrentLevel();
            if (x < l.getLeftLimit() || x + width > l.getRightLimit()) direction = !direction;
        }

        if(direction&&Main.getHarold().getX()+Main.getHarold().getWidth()<x||!direction&&Main.getHarold().getX()>x+width)direction=!direction;

        //Current sprite calculations
        if(damageCooldown>0){
            damageCooldown--;
            state=DAMAGE;
        }else if(state==DAMAGE){
            state=0;
        }

        doAttackCalc();
        if(state==ATTACK&&redAnimator.getCurrentFrameNum()==redAnimator.getFrames().length-1){
            redAnimator.setDelay(7);
            state=NORMAL;
            Attack.melee(this,1,5);
        }

        if(vX==0)redAnimator.setFrames(new ImageResource[]{ResourceHandler.getBossLoader().getRedMajorStill(direction)});
        else {
            doSpriteCalc();
        }
        redAnimator.update();
    }

    private void doSpriteCalc(){
        if(redAnimator.getDelay()==0)
        switch (state) {
            case NORMAL:
                redAnimator.setFrames(ResourceHandler.getBossLoader().getRedMajorWalk(direction));
                break;
            case READYING:
                redAnimator.setFrames(ResourceHandler.getBossLoader().getRedMajorReady(direction));
                break;
            case ATTACK:
                redAnimator.setFrames(ResourceHandler.getBossLoader().getRedMajorAttack(direction));
                break;
            case DAMAGE:
                redAnimator.setFrames(new ImageResource[]{ResourceHandler.getBossLoader().getRedMajorDamage(direction)});
                break;
        }

        if(doStartReady&&redAnimator.getCurrentFrameNum()==redAnimator.getFrames().length-1){
            redMajor=redAnimator.getCurrentFrame();
            redAnimator.setDelay(60);
            redAnimator.setFps(7);
            doStartReady=false;
            state=NORMAL;
        }

        if(state==READYING&&redAnimator.getCurrentFrameNum()==redAnimator.getFrames().length-1){
            redMajor=redAnimator.getCurrentFrame();
            redAnimator.setDelay(1);
            state=ATTACK;
            redAnimator.setCurrentFrame(0);
        }
        redMajor = redAnimator.getCurrentFrame();
    }

    private void doAttackCalc(){
        if(attackCooldown>0){
            attackCooldown--;
            return;
        }
        if(Main.getHarold().getY()>y+height||Main.getHarold().getY()+Main.getHarold().getWidth()<y){
            state=NORMAL;
            return;
        }
        if(direction){
            if(Main.getHarold().getX()>=x&&Main.getHarold().getX()<=x+width+5){
                state=READYING;
                redAnimator.setCurrentFrame(0);
                attackCooldown=100;
            }
        }else{
            if(Main.getHarold().getX()+Main.getHarold().getWidth()<=x+width&&Main.getHarold().getWidth()+Main.getHarold().getX()>=x-5){
                state=READYING;
                redAnimator.setCurrentFrame(0);
                attackCooldown=100;
            }
        }
    }

    @Override
    public void render() {
        if(redMajor==null)return;
        width= Graphics.convertToWorldWidth(redMajor.getTexture().getWidth());
        height=Graphics.convertToWorldHeight(redMajor.getTexture().getHeight());
        hitbox.updateBounds(x,y,width,height);
        if(damageTakenFrame>0){
            Graphics.setDrawColor(1,0,0,1);//Set damage color if needed
        }
        else Graphics.setDrawColor(1,1,1,1);
        Graphics.drawImage(redMajor,x,y);
        Graphics.setDrawColor(1,1,1,1);
    }

    @Override
    public void reset() {
        direction=false;
        displayName="Red Major";
        redAnimator=new Animator(ResourceHandler.getBossLoader().getRedMajorReady(false),1);
        redMajor=redAnimator.getCurrentFrame();
        health=8;
        maxHealth=8;
        state=READYING;
        doStartReady=true;
        startFight=false;
        redAnimator.setDelay(60);
        x=75;
        y=7;
        invincible=false;
    }

    @Override
    public String toString() {
        return "Red Major @ "+x+","+y;
    }

    public BossBar getBossBar() {
        return bossBar;
    }

    @Override
    public void handleDeath() {
        bossBar.update();
    }

    @Override
    public void doDamage(Entity attacker, int damage) {
        if(health>1)super.doDamage(attacker, damage);
        else{
            invincible=true;
            state=DEATH;
            redAnimator.setDelay(60);
            redAnimator.setFrames(new ImageResource[]{ResourceHandler.getBossLoader().getRedMajorDeath(direction)});
            redMajor=redAnimator.getCurrentFrame();
        }
    }
}

File: src/main/java/org/entities/aggressive/RedMajor.kt
package org.entities.aggressive

import org.engine.Main
import org.entities.Autonomous
import org.entities.Entity
import org.entities.SmartRectangle
import org.graphics.Animator
import org.graphics.BossBar
import org.graphics.Graphics
import org.level.LevelController.currentLevel
import org.loader.ImageResource
import org.loader.ResourceHandler
import org.world.Attack.melee
import org.world.HeightMap
import org.world.World.gravity

class RedMajor : Autonomous(5, 75f, 7f) {
    private val NORMAL = 0
    private val READYING = 1
    private val ATTACK = 2
    private val DAMAGE = 3
    private val DEATH = 4
    private var redAnimator: Animator? = null
    private var redMajor: ImageResource? = null
    private val hitbox = SmartRectangle(x, y, width, height)
    val bossBar = BossBar(this)
    private var doStartReady = true
    private var startFight = false
    fun setStartFight(startFight: Boolean) {
        this.startFight = startFight
    }

    override fun update(deltaTime: Float) {
        bossBar.update()
        if (state == 4) {
            if (redAnimator!!.delay > 0) redAnimator!!.update() else health = 0
        }
        if (!startFight || state == 4) return
        val heightReturn = HeightMap.onGround(hitbox)
        //Determines movement and knockback
        if (damageTakenFrame == 0f) {
            vX = if (direction) {
                10f
            } else {
                -10f
            }
        } else {
            vX = if (direction && !attackerBehind || !direction && attackerBehind) -2f else 2f
            damageTakenFrame--
        }

        //Calculates the y-value and velocity
        y += vY
        vY -= gravity
        if (heightReturn.isOnGround && vY < 0) {
            y = heightReturn.groundLevel
            vY = 0f
        }

        //Calculates the x-value and velocity
        if (state != READYING && redAnimator!!.delay == 0L) {
            x += vX * deltaTime
            val l = currentLevel
            if (x < l.leftLimit || x + width > l.rightLimit) direction = !direction
        }
        if (direction && Main.getHarold().x + Main.getHarold().width < x || !direction && Main.getHarold().x > x + width) direction =
            !direction

        //Current sprite calculations
        if (damageCooldown > 0) {
            damageCooldown -= (100 * deltaTime).toInt()
            state = DAMAGE
        } else if (state == DAMAGE) {
            state = 0
        }
        doAttackCalc(deltaTime)
        if (state == ATTACK && redAnimator!!.onLastFrame()) {
            redAnimator!!.delay = 7
            state = NORMAL
            melee(this, 1, 5f)
        }
        if (vX == 0f) redAnimator!!.frames = arrayOf(ResourceHandler.getBossLoader().getRedMajorStill(direction)) else {
            doSpriteCalc()
        }
        redAnimator!!.update()
    }

    private fun doSpriteCalc() {
        if (redAnimator!!.delay == 0L) when (state) {
            NORMAL -> redAnimator!!.frames = ResourceHandler.getBossLoader().getRedMajorWalk(direction)
            READYING -> redAnimator!!.frames = ResourceHandler.getBossLoader().getRedMajorReady(direction)
            ATTACK -> redAnimator!!.frames = ResourceHandler.getBossLoader().getRedMajorAttack(direction)
            DAMAGE -> redAnimator!!.frames = arrayOf(ResourceHandler.getBossLoader().getRedMajorDamage(direction))
        }
        if (doStartReady && redAnimator!!.onLastFrame()) {
            redMajor = redAnimator!!.currentFrame
            redAnimator!!.delay = 60
            redAnimator!!.fps = 7
            doStartReady = false
            state = NORMAL
        }
        if (state == READYING && redAnimator!!.onLastFrame()) {
            redMajor = redAnimator!!.currentFrame
            redAnimator!!.delay = 1
            state = ATTACK
            redAnimator!!.setCurrentFrame(0)
        }
        redMajor = redAnimator!!.currentFrame
    }

    private fun doAttackCalc(deltaTime: Float) {
        if (attackCooldown > 0) {
            attackCooldown -= 100 * deltaTime
            return
        }
        if (Main.getHarold().y > y + height || Main.getHarold().y + Main.getHarold().width < y) {
            state = NORMAL
            return
        }
        if (direction) {
            if (Main.getHarold().x >= x && Main.getHarold().x <= x + width + 5) {
                state = READYING
                redAnimator!!.setCurrentFrame(0)
                attackCooldown = 100f
            }
        } else {
            if (Main.getHarold().x + Main.getHarold().width <= x + width && Main.getHarold().width + Main.getHarold().x >= x - 5) {
                state = READYING
                redAnimator!!.setCurrentFrame(0)
                attackCooldown = 100f
            }
        }
    }

    override fun render() {
        if (redMajor == null) return
        width = Graphics.toWorldWidth(redMajor!!.texture.width.toFloat())
        height = Graphics.toWorldHeight(redMajor!!.texture.height.toFloat())
        hitbox.updateBounds(x, y, width, height)
        if (damageTakenFrame > 0) {
            Graphics.setDrawColor(1f, 0f, 0f, 1f) //Set damage color if needed
        } else Graphics.setDrawColor(1f, 1f, 1f, 1f)
        Graphics.drawImage(redMajor!!, x, y)
        Graphics.setDrawColor(1f, 1f, 1f, 1f)
    }

    override fun reset() {
        direction = false
        displayName = "Red Major"
        redAnimator = Animator(ResourceHandler.getBossLoader().getRedMajorReady(false), 1)
        redMajor = redAnimator!!.currentFrame
        health = 8
        maxHealth = 8
        state = READYING
        doStartReady = true
        startFight = false
        redAnimator!!.delay = 60
        x = 75f
        y = 7f
        invincible = false
    }

    override fun handleDeath() {
        bossBar.update()
    }

    override fun doDamage(attacker: Entity, damage: Int) {
        if (health > 1) super.doDamage(attacker, damage) else {
            invincible = true
            state = DEATH
            redAnimator!!.delay = 60
            redAnimator!!.frames = arrayOf(ResourceHandler.getBossLoader().getRedMajorDeath(direction))
            redMajor = redAnimator!!.currentFrame
        }
    }

    init {
        reset()
    }
}